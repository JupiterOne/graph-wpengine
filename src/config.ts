import {
  IntegrationExecutionContext,
  IntegrationValidationError,
  IntegrationInstanceConfigFieldMap,
  IntegrationInstanceConfig,
} from '@jupiterone/integration-sdk-core';
import { createAPIClient } from './client';

export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  wpEngineUsername: {
    type: 'string',
  },
  wpEnginePassword: {
    type: 'string',
    mask: true,
  },
};

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  /**
   * The provider API client ID used to authenticate requests.
   */
  wpEngineUsername: string;

  /**
   * The provider API client secret used to authenticate requests.
   */
  wpEnginePassword: string;
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationConfig>,
) {
  const { config } = context.instance;

  if (!config.wpEngineUsername || !config.wpEnginePassword) {
    throw new IntegrationValidationError(
      'Config requires all of {wpEngineUsername, wpEnginePassword}',
    );
  }

  const apiClient = createAPIClient(config);
  await apiClient.verifyAuthentication();
}
