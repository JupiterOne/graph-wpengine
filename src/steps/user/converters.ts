import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { Entities } from '../constants';
import { WpEngineUser } from '../../types';

export function createUserEntity(data: WpEngineUser) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.USER._class,
        _type: Entities.USER._type,
        // We always want to have "descriptive" keys
        // Also, if later, you need to find users (jobState.findEntity())
        // you'll need to find them by their keys, so you should
        // export and turn this into function similar to getAccountKey
        _key: `wp_engine_user:${data.id}`,
        name: `${data.first_name} ${data.last_name}`,
        username: data.email,
        email: data.email,
      },
    },
  });
}
