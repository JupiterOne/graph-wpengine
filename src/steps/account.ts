import {
  // createIntegrationEntity,
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

import { IntegrationConfig } from '../config';
import { createAPIClient } from '../client';
import { entities, steps } from '../constants';

export async function fetchAccounts({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  await apiClient.iterateAccounts((account) => {
    console.log(account);
  });
}

export const accountSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: steps.FETCH_ACCOUNT,
    name: 'Fetch Account Details',
    entities: [entities.ACCOUNT],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchAccounts,
  },
];
