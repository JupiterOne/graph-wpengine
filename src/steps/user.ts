import {
  createIntegrationEntity,
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

import { IntegrationConfig } from '../config';
import { createAPIClient } from '../client';
import { entities, steps } from '../constants';

export const USER_ENTITY_KEY = 'entity:user';

export async function fetchUserDetails({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  const user = await apiClient.getUser();

  const userEntity = await jobState.addEntity(
    createIntegrationEntity({
      entityData: {
        source: user,
        assign: {
          _key: user.id,
          _type: entities.USER._type,
          _class: entities.USER._class,
          name: `${user.first_name} ${user.last_name}`,
          username: user.email,
          email: user.email,
        },
      },
    }),
  );

  await jobState.setData(USER_ENTITY_KEY, userEntity);
}

export const userSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: steps.FETCH_USER,
    name: 'Fetch User Details',
    entities: [entities.USER],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchUserDetails,
  },
];
