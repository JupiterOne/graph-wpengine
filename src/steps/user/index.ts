import {
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

import { IntegrationConfig } from '../../config';
import { createAPIClient } from '../../client';
import { USER_ENTITY_KEY, Entities, IntegrationSteps } from '../constants';
import { createUserEntity } from './converters';

export async function fetchUser({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  const user = await apiClient.getUser();

  const userEntity = createUserEntity(user);
  await jobState.addEntity(userEntity);
  await jobState.setData(USER_ENTITY_KEY, userEntity);
}

export const userSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: IntegrationSteps.USER,
    name: 'Fetch User',
    entities: [Entities.USER],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchUser,
  },
];
