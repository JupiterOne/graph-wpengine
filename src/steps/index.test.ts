import { createMockStepExecutionContext } from '@jupiterone/integration-sdk-testing';

import { IntegrationConfig } from '../config';
import { fetchUserDetails } from './user';
import { fetchAccounts } from './account';
import { integrationConfig } from '../../test/config';

test('should collect data', async () => {
  const context = createMockStepExecutionContext<IntegrationConfig>({
    instanceConfig: integrationConfig,
  });

  // Simulates dependency graph execution.
  // See https://github.com/JupiterOne/sdk/issues/262.
  await fetchUserDetails(context);
  await fetchAccounts(context);

  // Review snapshot, failure is a regression
  expect({
    numCollectedEntities: context.jobState.collectedEntities.length,
    numCollectedRelationships: context.jobState.collectedRelationships.length,
    collectedEntities: context.jobState.collectedEntities,
    collectedRelationships: context.jobState.collectedRelationships,
    encounteredTypes: context.jobState.encounteredTypes,
  }).toMatchSnapshot();

  const user = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('User'),
  );
  expect(user.length).toBeGreaterThan(0);
  expect(user).toMatchGraphObjectSchema({
    _class: ['User'],
  });

  const accounts = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('Account'),
  );
  expect(accounts.length).toBeGreaterThan(0);
  expect(accounts).toMatchGraphObjectSchema({
    _class: ['Account'],
  });
});
