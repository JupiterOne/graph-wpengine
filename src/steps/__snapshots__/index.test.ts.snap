// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "3937869a-09a0-4198-a1c7-d70c451c24fc",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "email": "stefan@creativice.com",
            "first_name": "Stefan",
            "id": "3937869a-09a0-4198-a1c7-d70c451c24fc",
            "last_name": "Virag",
          },
        },
      ],
      "_type": "wp_engine_user",
      "createdOn": undefined,
      "displayName": "Stefan Virag",
      "email": "stefan@creativice.com",
      "id": "3937869a-09a0-4198-a1c7-d70c451c24fc",
      "name": "Stefan Virag",
      "username": "stefan@creativice.com",
    },
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "wp_engine_account:77815c5d-e25c-41a3-ba2d-29f8ceb82ac1",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "77815c5d-e25c-41a3-ba2d-29f8ceb82ac1",
            "name": "j1inc",
          },
        },
      ],
      "_type": "wp_engine_account",
      "createdOn": undefined,
      "displayName": "j1inc",
      "id": "77815c5d-e25c-41a3-ba2d-29f8ceb82ac1",
      "name": "j1inc",
    },
  ],
  "collectedRelationships": Array [],
  "encounteredTypes": Array [
    "wp_engine_user",
    "wp_engine_account",
  ],
  "numCollectedEntities": 2,
  "numCollectedRelationships": 0,
}
`;
